### PR TITLE
analyzer: Run package managers in parallel

### PR DESCRIPTION
**Pull request by mnonnenmacher**
This is an experiment to execute different package managers in parallel
to check how it influences the analyzer performance.

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/1615)

### CODE REVIEW

File: analyzer/build.gradle.kts [hunk 1]
```diff
@@ -36,4 +37,6 @@ dependencies {
     implementation("org.apache.maven.resolver:maven-resolver-transport-wagon:$mavenResolverVersion")
 
     implementation("org.gradle:gradle-tooling-api:${gradle.gradleVersion}")
+
```
Comment (sschuberth) [inline analyzer/build.gradle.kts]: Nit: I don't think we need a blank line here.

File: analyzer/src/main/kotlin/Analyzer.kt [hunk 2]
```diff
@@ -144,11 +125,56 @@ class Analyzer(private val config: AnalyzerConfiguration) {
...
 
         return OrtResult(repository, run)
     }
 
+    @ObsoleteCoroutinesApi
+    private suspend fun analyzeInParallel(
+        managedFiles: Map<PackageManager, List<File>>,
+        packageCurationsFile: File?
+    ): AnalyzerResult {
+        val dispatcher = newFixedThreadPoolContext(5, "storage")
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/Analyzer.kt]: The name should be "analyse" or similar.

File: analyzer/src/funTest/kotlin/SbtTest.kt [hunk 3]
```diff
@@ -31,6 +31,9 @@ import io.kotlintest.specs.StringSpec
 
 import java.io.File
 
+import kotlinx.coroutines.ObsoleteCoroutinesApi
```
Comment (sschuberth) [inline analyzer/src/funTest/kotlin/SbtTest.kt]: Left-over?

File: analyzer/src/funTest/kotlin/vcs/GitRepoTest.kt [hunk 4]
```diff
@@ -36,10 +36,13 @@ import io.kotlintest.specs.StringSpec
 
 import java.io.File
 
+import kotlinx.coroutines.ObsoleteCoroutinesApi
+
 private const val REPO_URL = "https://github.com/heremaps/oss-review-toolkit-test-data-git-repo"
 private const val REPO_REV = "f00ec4cbb670b49a156fd95d29e8fd148d931ba9"
 private const val REPO_MANIFEST = "manifest.xml"
 
+@ObsoleteCoroutinesApi
```
Comment (sschuberth) [inline analyzer/src/funTest/kotlin/vcs/GitRepoTest.kt]: Left-over?

File: analyzer/src/main/kotlin/Analyzer.kt [hunk 5]
```diff
@@ -41,13 +42,21 @@ import com.here.ort.utils.realFile
...
+import kotlinx.coroutines.withContext
+
 const val TOOL_NAME = "analyzer"
 const val HTTP_CACHE_PATH = "$TOOL_NAME/cache/http"
 
 /**
  * The class to run the analysis. The signatures of public functions in this class define the library API.
  */
 class Analyzer(private val config: AnalyzerConfiguration) {
+    @ObsoleteCoroutinesApi
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/Analyzer.kt]: Can we please repeat the comment from `scanPackages` about the obsolete API here?

Comment (sschuberth) [reply analyzer/src/main/kotlin/Analyzer.kt]: Actually, is the annotation required here, or is it sufficient to have it below?

**Comment by sschuberth** [comment]
Just FYI, in relation to this PR, see this article about [Working with Kotlin Coroutines](https://link.medium.com/nVQMuPKm3X).

**Comment by mnonnenmacher** [comment]
A short update: Test show that this change can significantly improve analyzer performance if multiple package managers are used in one project. The downside is that the logs become less readable.
However, as I'm currently focusing on other topics like the PostgreSQL backend it will take some weeks before I find time to continue work on this, so if anyone wants to take over just let me know.

**Comment by sschuberth** [comment]
> The downside is that the logs become less readable.

That situation probably could be improved by using different loggers per package manager and adjusting our logger pattern to include the logger name.

File: analyzer/build.gradle.kts (hunk 1 above)
Comment (mnonnenmacher) [reply analyzer/build.gradle.kts]: Done.

File: analyzer/src/main/kotlin/Analyzer.kt (hunk 2 above)
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/Analyzer.kt]: Done.

File: analyzer/src/funTest/kotlin/SbtTest.kt (hunk 3 above)
Comment (mnonnenmacher) [reply analyzer/src/funTest/kotlin/SbtTest.kt]: I'm using a different way to create the dispatcher now, not using the obsolete API anymore.

File: analyzer/src/main/kotlin/Analyzer.kt (hunk 5 above)
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/Analyzer.kt]: I'm using a different way to create the dispatcher now, not using the obsolete API anymore. But the annotation was actually required in all places I used it.

File: analyzer/src/funTest/kotlin/SbtTest.kt (hunk 3 above)
Comment (sschuberth) [reply analyzer/src/funTest/kotlin/SbtTest.kt]: Sounds great! Are you planning to also update the existing parallel scanner code accordingly (in a new PR, of course)?

**Comment by sschuberth** [comment]
@mnonnenmacher, you've removed the "WIP" from the title, but the PR is still marked as a draft. I guess it's ready for review now?

File: analyzer/src/funTest/kotlin/SbtTest.kt (hunk 3 above)
Comment (sschuberth) [reply analyzer/src/funTest/kotlin/SbtTest.kt]: Ah, I just saw your `scanner-no-obsolete-api` branch, perfect :-)

**Comment by mnonnenmacher** [comment]
> @mnonnenmacher, you've removed the "WIP" from the title, but the PR is still marked as a draft. I guess it's ready for review now?

Yes it is, also see: https://github.com/heremaps/oss-review-toolkit/pull/1825

**Pull request merged**